src/Microsoft.AspNet.Server.Kestrel/Http/ListenerSecondary.cs

```diff
@@ -29,95 +33,117 @@ protected ListenerSecondary(ServiceContext serviceContext) : base(serviceContext
             KestrelThread thread,
             RequestDelegate application)
         {
+            _pipeName = pipeName;
             ServerAddress = address;
             Thread = thread;
             Application = application;
 
             DispatchPipe = new UvPipeHandle(Log);
 
-            var tcs = new TaskCompletionSource<int>();
-            Thread.Post(_ =>
+            var tcs = new TaskCompletionSource<int>(this);
+            Thread.Post(tcs2 => StartCallback(tcs2), tcs);
+            return tcs.Task;
+        }
+
+        private static void StartCallback(TaskCompletionSource<int> tcs)
+        {
+            var listener = (ListenerSecondary)tcs.Task.AsyncState;
+            listener.StartedCallback(tcs);
+        }
+
+        private void StartedCallback(TaskCompletionSource<int> tcs)
+        {
+            try
             {
-                try
-                {
-                    DispatchPipe.Init(Thread.Loop, true);
-                    var connect = new UvConnectRequest(Log);
-                    connect.Init(Thread.Loop);
-                    connect.Connect(
-                        DispatchPipe,
-                        pipeName,
-                        (connect2, status, error, state) =>
+                DispatchPipe.Init(Thread.Loop, true);
+                var connect = new UvConnectRequest(Log);
+                connect.Init(Thread.Loop);
+                connect.Connect(
+                    DispatchPipe,
+                    _pipeName,
+                    (connect2, status, error, state) => ConnectCallback(connect2, status, error, (TaskCompletionSource<int>)state),
+                    tcs);
+            }
+            catch (Exception ex)
+            {
+                DispatchPipe.Dispose();
+                tcs.SetException(ex);
+            }
+        }
+
+        private static void ConnectCallback(UvConnectRequest connect, int status, Exception error, TaskCompletionSource<int> tcs)
+        {
+            var listener = (ListenerSecondary)tcs.Task.AsyncState;
+            listener.ConnectedCallback(connect, status, error, tcs);
+        }
+
+        private void ConnectedCallback(UvConnectRequest connect, int status, Exception error, TaskCompletionSource<int> tcs)
+        {
+            connect.Dispose();
+            if (error != null)
+            {
+                tcs.SetException(error);
+                return;
+            }
+
+            try
+            {
+                _ptr = Marshal.AllocHGlobal(4);
+                _buf = Thread.Loop.Libuv.buf_init(_ptr, 4);
```

Comment:
I would just initialize these fields in the ctor. Then we can have only one call to `Marshal.FreeHGlobal` in Dispose.